src/QsCompiler/CompilerOptimizations/Utils.fs

```diff
@@ -238,3 +130,82 @@ let rec internal takeWhilePlus1 (f: 'A -> bool) (l : list<'A>) =
         else [], Some first
     | [] -> [], None
 
+
+/// Returns all the "bottom-level" statements in the current scope.
+/// Bottom-level statements are defined as those that don't contain their own scopes.
+/// This function recurses through all the subscopes of the current scope.
+let rec private findAllBaseStatements (scope: QsScope): seq<QsStatementKind> =
+    scope.Statements |> Seq.collect (fun s ->
+        match s.Statement with
+        | QsConditionalStatement cond ->
+            seq {
+                for _, b in cond.ConditionalBlocks do yield b.Body
+                match cond.Default with
+                | Value b -> yield b.Body
+                | _ -> ()
+            } |> Seq.collect findAllBaseStatements
+        | QsForStatement {Body = scope}
+        | QsWhileStatement {Body = scope}
+        | QsQubitScope {Body = scope}
+        | QsScopeStatement {Body = scope} -> findAllBaseStatements scope
+        | QsRepeatStatement {RepeatBlock = scope1; FixupBlock = scope2} ->
+            Seq.concat [findAllBaseStatements scope1.Body; findAllBaseStatements scope2.Body]
+        | x -> Seq.singleton x
+    )
+
+/// Returns whether this scope contains any return statement
+let rec internal hasReturnStatement (scope: QsScope): bool =
+    scope |> findAllBaseStatements |> Seq.exists (function QsReturnStatement _ -> true | _ -> false)
+
+/// Returns the number of "bottom-level" statements in this scope
+let rec internal scopeLength (scope: QsScope): int =
+    scope |> findAllBaseStatements |> Seq.length
+
+
+/// Tries to inline the given expression, checking whether it's valid to do so.
+/// Returns the name of the callable and the modified scope after inlining.
+let internal tryInline (state: TransformationState) (ex: TypedExpression) =
+    maybe {
+        let! expr, arg =
+            match ex.Expression with
+            | CallLikeExpression ({Expression = expr}, arg) -> Some (expr, arg)
+            | _ -> None
+        let! qualName, specKind =
+            match expr with
```

Comment:
You should probably make it recursive, so it handles an arbitrary list of preceding modifiers (functors).

Reply:
Yeah, I'll fix that in a future PR